apps/studio/src/lib/db/clients/duckdb.ts

```diff
@@ -0,0 +1,944 @@
+        log.error(error);
+        throw error;
+      }
+    }
+
+    return options.multiple ? results : results[0];
+  }
+
+  async getPrimaryKey(table: string, schema: string): Promise<string> {
+    const keys = await this.getPrimaryKeys(table, schema);
+    return keys.length === 1 ? keys[0].columnName : null;
+  }
+
+  async getPrimaryKeys(
+    table: string,
+    schema: string
+  ): Promise<PrimaryKeyColumn[]> {
+    const keys: PrimaryKeyColumn[] = [];
+
+    const sql = `
+      SELECT
+        constraint_column_indexes,
+        constraint_column_names
+      FROM duckdb_constraints
+      WHERE schema_name = ?
+        AND table_name = ?
+        AND constraint_type = 'PRIMARY KEY'
+    `;
+
+    const { data } = await this.driverExecuteSingle(sql, {
+      params: [schema, table],
+    });
+
+    data.forEach((row: any) => {
+      const columnNames: string[] = row.constraint_column_names;
+      columnNames.forEach((name, idx) => {
+        keys.push({
+          columnName: name,
+          position: Number(row.constraint_column_indexes[idx]),
+        });
+      });
+    });
+
+    return keys;
+  }
+
+  async listCollations(): Promise<string[]> {
+    const result = await this.driverExecuteSingle("PRAGMA collations");
+    return result.data.map((row) => row.collname);
+  }
+
+  async listCharsets(): Promise<string[]> {
+    // No charsets in the docs
+    return [];
+  }
+
+  async getDefaultCharset(): Promise<string> {
+    // No charsets in the docs
+    return "";
+  }
+
+  /**
+   * @param {string} databaseName - A path to a file
+   */
+  async createDatabase(
+    databaseName: string,
+    _charset: string,
+    _collation: string
+  ): Promise<void> {
+    const database = await Database.create(databaseName);
+    await database.close();
+  }
+
+  createDatabaseSQL(): string {
+    throw new Error("Method not implemented.");
+  }
+
+  async getTableCreateScript(table: string, schema: string): Promise<string> {
+    const { data } = await this.driverExecuteSingle(
+      `
+        SELECT sql
+        FROM duckdb_tables
+        WHERE table_name = ?
+        AND schema_name = ?
+      `,
+      { params: [table, schema] }
+    );
+    return data[0].sql;
+  }
+
+  async getViewCreateScript(view: string, schema: string): Promise<string[]> {
+    const { data } = await this.driverExecuteSingle(
+      `
+        SELECT sql
+        FROM duckdb_views
+        WHERE view_name = ?
+        AND schema_name = ?
+      `,
+      { params: [view, schema] }
+    );
+    return data.map((row) => row.sql);
+  }
+
+  async getRoutineCreateScript(
+    _routine: string,
+    _type: string,
+    _schema: string
+  ): Promise<string[]> {
+    return [];
+  }
+
+  applyChangesSql(changes: TableChanges): string {
+    return applyChangesSql(changes, this.knex);
+  }
+
+  async applyChanges(changes: TableChanges): Promise<TableUpdateResult[]> {
+    let results = [];
+
+    await this.driverExecuteSingle("BEGIN TRANSACTION");
```

Comment:
Does duckdb only have one constant connection? If not, there's a danger of these queries (`begin transaction`, the update stuff, etc) being run on different connections. If it is truly just like sqlite though, then we don't have to worry about it.

Reply:
DuckDB can have multiple connections, but it won't allow write actions if there are multiple connections. So it's impossible to do transactions in that situation. We need to only have one connection to be able to both read and write.